engine/resources/hetzner.go

```diff
@@ -0,0 +1,710 @@
+
+	// ServerRescueMode specifies the image used when enabling rescue mode.
+	// The supported images are "linux32", "linux64" and "freebsd64".
+	// Alternatively, use "disabled" to disable rescue mode (default).
+	ServerRescueMode string `lang:"serverrescuemode"`
+	// ServerRescueSSHKeys is used to select a subset of keys that
+	// should be available for rescue mode operations over ssh.
+	// From all ssh keys known to the project client, choose a subset of
+	// keys by name, as an array of strings. New keys need to be added
+	// manually via the cloud console web api for now.
+	// If a keyname is unknown to the client, it is ignored.
+	ServerRescueSSHKeys []string `lang:"serverrescuekeys"`
+
+	// Full client details for interaction with the hcloud-go API.
+	client *hcloud.Client
+	// Snapshot of the server instance, for checkapply operations.
+	server *hcloud.Server
+	// Snapshot of the serverCreateOpts struct generated with hcloud.
+	serverconfig hcloud.ServerCreateOpts
+	// hcloud-formatted array of ssh keys for use with rescue mode.
+	rescuekeys []*hcloud.SSHKey
+
+	// Internal flag to simplify rescue mode operations in checkapply.
+	// False if ServerRescueMode="disabled", true otherwise.
+	rescueEnable bool
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *HetznerRes) Default() engine.Res {
+	return &HetznerRes{
+		WaitInterval:     WaitIntervalLimit,
+		ServerRescueMode: "disabled",
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *HetznerRes) Validate() error {
+
+	// validate token
+	if obj.APIToken == "" {
+		return fmt.Errorf("empty token string")
+	}
+	// validate time intervals
+	if obj.MetaParams().Poll < HetznerPollLimit {
+		return fmt.Errorf("invalid polling interval (min = %d s)", HetznerPollLimit)
+	}
+	if obj.WaitInterval < WaitIntervalLimit {
+		return fmt.Errorf("invalid wait interval (min = %d)", WaitIntervalLimit)
+	}
+	// validate state
+	switch obj.State {
+	case StateRunning, StateOff, StateAbsent:
+		// valid state options
+	default:
+		return fmt.Errorf("invalid state")
+	}
+	// validate rescue mode parameters
+	switch obj.ServerRescueMode {
+	case ServerRescueTypeLinux32, ServerRescueTypeLinux64, ServerRescueTypeFreeBSD64:
+		// valid options for rescue mode image
+	case ServerRescueDisabled:
+		// valid option to disable rescue mode
+	default:
+		return fmt.Errorf("invalid serverrescuemode")
+	}
+
+	return nil
+}
+
+// Init runs some startup code for this resource.
+func (obj *HetznerRes) Init(init *engine.Init) error {
+
+	// save init struct
+	obj.init = init
+
+	// initialize hcloud client
+	token := obj.APIToken
+	obj.client = hcloud.NewClient(hcloud.WithToken(token))
+
+	// convert ServerRescueMode to rescueEnable flag
+	obj.rescueEnable = (obj.ServerRescueMode != ServerRescueDisabled)
+
+	return nil
+}
+
+// Close runs some cleanup code for this resource.
+// TODO: clear the root password info if rescue mode was enabled?
```

Comment:
I will clarify that in the code comments. Right now I am not even reading the root password when enabling rescue mode, so that you can only connect to the server over SSH. The ssh key implementation seems to work as intended, so there is no urgent security threat at the moment. If there is ever a desire to add password support though, I already added those TODO's in the relevant places to remind me (or other devs) what to pay attention to.